compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs

```diff
@@ -241,14 +200,118 @@ safetyStep = \case
+isTypeClassDictionary DefValue{..}
+    = T.isPrefixOf "$f" (unExprValName (fst dvalBinder)) -- generic dictionary
+    || T.isPrefixOf "$d" (unExprValName (fst dvalBinder)) -- specialized dictionary
 
-simplifyExpr :: Expr -> Expr
-simplifyExpr = fst . cata go
+isTypeClassProjection :: DefValue -> Bool
+isTypeClassProjection DefValue{..} = go dvalBody
   where
+    go :: Expr -> Bool
+    go (ETyLam _ e) = go e
+    go (ETmLam _ e) = go e
+    go (ETmApp e _) = go e
+    go (EStructProj _ _) = True
+    go _ = False
+
+typeclassStep :: World -> ExprF TypeClassInfo -> TypeClassInfo
+typeclassStep world = \case
+    EValF x ->
+        case lookupValue x world of
+            Left _ -> TCNeither
+            Right dv
+                | isTypeClassProjection dv -> TCProjection (dvalBody dv)
+                | isTypeClassDictionary dv -> TCDictionary (dvalBody dv)
+                | otherwise -> TCNeither
+
+    ETyAppF tci ty ->
+        case tci of
+            TCProjection (ETyLam (x,_) e) ->
+                TCProjection (substExpr (typeSubst x ty) e)
+            TCDictionary (ETyLam (x,_) e) ->
+                TCDictionary (substExpr (typeSubst x ty) e)
+            _ ->
+                TCNeither
+
+    _ -> TCNeither
+
+
+-- -- | Should I apply this argument immediately in a lambda?
+-- shouldApply :: Expr -> Bool
+-- shouldApply = \case
+--     EBuiltin _ -> True
+-- --    EStructCon _ -> True
+--     _ -> False
+
+infoStep :: World -> ExprF Info -> Info
+infoStep world e = Info
+    (freeVarsStep (fmap freeVars e))
+    (safetyStep (fmap safety e))
+    (typeclassStep world (fmap tcinfo e))
+
+-- | Try to get the actual field value from the body of
+-- a typeclass projection function, after substitution of the
+-- dictionary function inside.
+getProjectedTypeclassField :: World -> Expr -> Maybe Expr
+getProjectedTypeclassField world = \case
+    EStructProj f e -> do
+        EStructCon fs <- getTypeClassDictionary world e
+        lookup f fs
+
+    ETmApp e (EBuiltin BEUnit) -> do
+        ETmLam (x,_) e' <- getProjectedTypeclassField world e
+        Just (substExpr (exprSubst x (EBuiltin BEUnit)) e')
+
+    _ ->
+        Nothing
+
+-- | Try to get typeclass dictionary from the body of
+-- a typeclass dictionary function, after substitution.
+-- This is made complicated by GHC's specializer, which
+-- introduces a level of indirection. That's why we need
+-- to inline dictionary functions and beta-reduce.
+getTypeClassDictionary :: World -> Expr -> Maybe Expr
+getTypeClassDictionary world = \case
+    e@(EStructCon _) ->
+        Just e
+
+    EVal x
+        | Right dv <- lookupValue x world
+        , isTypeClassDictionary dv
+        -> do
+            Just (dvalBody dv)
+
+    ETyApp e t -> do
+        ETyLam (x,_) e' <- getTypeClassDictionary world e
+        Just (substExpr (typeSubst x t) e')
+
+    ETmApp e1 e2 -> do
+        ETmLam (x,_) e1' <- getTypeClassDictionary world e1
+        Just (substExpr (exprSubst x e2) e1')
```

Comment:
Aren't we running danger to duplicate code here? Should we avoid this with a `let` binding when necessary?

Reply:
I don't think so. This is here because of typeclass specialization.

Introducing a let would make the optimization a lot more complicated -- can't apply typeclass projection on a let directly.